SH-2A + FPU Support

loudinthecloud opened the pull request:
Hi @VGKintsugi,

As per @mumbel [suggestion](https://github.com/VGKintsugi/Ghidra-SegaSaturn-Processor/issues/3), here is our PR.

Thanks!

mumbel commented on Ghidra/Processors/SuperH/data/languages/superh.sinc:
```diff
@@ -257,24 +361,248 @@ target00_11: target is sdisp_00_11 [ target = (sdisp_00_11 << 1) + inst_start +
+    { export *[ram]:4 value; }
+
+imm20s: "#"value is l_simm20_20_23 & l_simm20_00_16
+    [ value = ((((l_simm20_20_23 << 16) | l_simm20_00_16) << 8)); ]
```
i read this as:  `value = (((l_simm20_20_23 << 16) | l_simm20_00_16)) << 8;` in the manual

loudinthecloud replied:
Hey, wow that was confusing, there are 2 extra spurious brackets but it seems to be semantically correct :)

mumbel replied:
Yup, sorry, guess the extras just threw me off

mumbel commented on Ghidra/Processors/SuperH/data/languages/superh.sinc:
```diff
@@ -257,24 +361,248 @@ target00_11: target is sdisp_00_11 [ target = (sdisp_00_11 << 1) + inst_start +
+}
+
+# MOVRT Rn                      0000nnnn00111001    ~ T → Rn
+:movrt rn_08_11 is opcode_12_15=0b0000 & rn_08_11 & opcode_00_07=0b00111001
```
`rn_08_11 = zext($(T_FLAG) == 0);` instead I think

mumbel commented on Ghidra/Processors/SuperH/data/languages/superh.sinc:
```diff
@@ -412,6 +738,60 @@ target00_11: target is sdisp_00_11 [ target = (sdisp_00_11 << 1) + inst_start +
 
+@if SH_VERSION == "2A"
+
+:clips.b rn_08_11 is opcode_12_15=0b0100 & rn_08_11 & opcode_00_07=0b10010001
```
this looks super weird since `goto <end>` would never hit as i read it.  It looks like it would be a signed comparison right?

edit havent compiled/tested this but written similar for my modules. you might consider an initial check of is saturated? no? goto inst_next and then do your saturation assignments, something like:
 
```
uppercheck = (rn_08_11 s> 0x7f)
lowercheck = (rn_08_11 s< -0x80)
if (!(uppercheck | lowercheck)) goto inst_next;
rn_08_11 = (0x7f * zext(uppercheck)) + sext(-0x80 * zext(lowercheck));
$(CS_FLAG)=1;
```
generally speaking the suggestions I've seen from the ghidra devs is to remove unnecessary goto of internal labels if possible (again maybe im incorrect in the comparisons, but it does look weird)

loudinthecloud replied:
Great catch! seems you are correct, the pseudo code in the manuals is weird indeed. I'll use your version.

mumbel commented on Ghidra/Processors/SuperH/data/languages/superh.sinc:
```diff
@@ -412,6 +738,60 @@ target00_11: target is sdisp_00_11 [ target = (sdisp_00_11 << 1) + inst_start +
+<end>
+}
+
+:clips.w rn_08_11 is opcode_12_15=0b0100 & rn_08_11 & opcode_00_07=0b10010101
```
same suggestion/comment about goto/weirdness as clips.b

mumbel commented on Ghidra/Processors/SuperH/data/languages/superh.sinc:
```diff
@@ -412,6 +738,60 @@ target00_11: target is sdisp_00_11 [ target = (sdisp_00_11 << 1) + inst_start +
+<end>
+}
+
+:clipu.b rn_08_11 is opcode_12_15=0b0100 & rn_08_11 & opcode_00_07=0b10000001
```
this can easily be simplified to remove the internal goto

mumbel commented on Ghidra/Processors/SuperH/data/languages/superh.sinc:
```diff
@@ -412,6 +738,60 @@ target00_11: target is sdisp_00_11 [ target = (sdisp_00_11 << 1) + inst_start +
+<end>
+}
+
+:clipu.w rn_08_11 is opcode_12_15=0b0100 & rn_08_11 & opcode_00_07=0b10000101
```
this can easily be simplified to remove the internal goto

mumbel commented on Ghidra/Processors/SuperH/data/languages/superh.sinc:
```diff
@@ -1408,3 +1927,453 @@ target00_11: target is sdisp_00_11 [ target = (sdisp_00_11 << 1) + inst_start +
+# TODO: FADD DRm, DRn             1111nnn0mmm00000 DRn + DRm → DRn
+
+# FCMP/EQ FRm, FRn          1111nnnnmmmm0100 (FRn=FRm)? 1:0 → T
+:fcmp"/eq" ffrm_04_07, ffrn_08_11   is fop_12_15=0b1111 & ffrn_08_11 & ffrm_04_07 & fop_00_03=0b0100
```
simplify to something like  `$(T_FLAG) = ffrn_08_11 == ffrm_04_07`

mumbel commented on Ghidra/Processors/SuperH/data/languages/superh.sinc:
```diff
@@ -1408,3 +1927,453 @@ target00_11: target is sdisp_00_11 [ target = (sdisp_00_11 << 1) + inst_start +
+# TODO: FCMP/EQ DRm, DRn          1111nnn0mmm00100 (DRn=DRm)? 1:0 → T
+
+# FCMP/GT FRm, FRn          1111nnnnmmmm0101 (FRn>FRm)? 1:0 → T
+:fcmp"/gt" ffrm_04_07, ffrn_08_11   is fop_12_15=0b1111 & ffrn_08_11 & ffrm_04_07 & fop_00_03=0b0101
```
simplify to remove goto

mumbel commented on Ghidra/Processors/SuperH/data/languages/superh.sinc:
```diff
@@ -1408,3 +1927,453 @@ target00_11: target is sdisp_00_11 [ target = (sdisp_00_11 << 1) + inst_start +
+}
+
+# BOR.B       #imm3, @(disp12,Rn)      0011nnnn0iii1001 0101dddddddddddd        (imm of (disp+ Rn)) | T → T
+:bor.b "#"l_imm3_20_22, @(l_disp_00_11, l_rn_24_27)
```
simplify goto

mumbel commented on Ghidra/Processors/SuperH/data/languages/superh.sinc:
```diff
@@ -1408,3 +1927,453 @@ target00_11: target is sdisp_00_11 [ target = (sdisp_00_11 << 1) + inst_start +
+}
+
+# BORNOT.B    #imm3, @(disp12,Rn)      0011nnnn0iii1001 1101dddddddddddd        ~ (imm of (disp+ Rn)) | T → T
+:bornot.b "#"l_imm3_20_22, @(l_disp_00_11, l_rn_24_27)
```
simplify goto

mumbel commented on Ghidra/Processors/SuperH/data/languages/superh.sinc:
```diff
@@ -1408,3 +1927,453 @@ target00_11: target is sdisp_00_11 [ target = (sdisp_00_11 << 1) + inst_start +
+}
+
+# BST.B       #imm3 ,@(disp12,Rn)      0011nnnn0iii1001 0010dddddddddddd        T → (imm of (disp+Rn))
+:bst.b      "#"l_imm3_20_22, @(l_disp_00_11, l_rn_24_27)
```
bst.b and bst shouldnt be to awful to get rid of the goto
`rn = ((b|ibit) * zext(tflag!=0)) + ((b & ~(ibit)) * zext(tflag==0))`

loudinthecloud commented on Ghidra/Processors/SuperH/data/languages/superh.sinc:
```diff
@@ -72,22 +177,21 @@ target00_11: target is sdisp_00_11 [ target = (sdisp_00_11 << 1) + inst_start +
 }
 
-:mova   @(disp,pc),r0  is r0 & pc & opcode_08_15=0b11000111 & disp_00_07  [disp = (disp_00_07 << 2); ]
+:mova   addr,r0  is r0 & opcode_08_15=0b11000111 & disp_00_07  [ addr = (inst_start + 4) & 0xFFFFFFFC + (disp_00_07 << 2); ]
```
In the case of `mova @(disp, pc), r0` the disassembly doesn't show the pointer value that is assigned to r0, is it possible to add a comment maybe?

mumbel replied:
Ah. I see. Writing on mobile so might have typos. Maybe try creating this and then using that in the instruction.  mova would simply be 

`:mova disppc4,r0 is r0 & opcode_08_15=0b11000111 & disppc4 { r0 = disppc4; }`
```
disppc2: @(disp,pc) is disp_00_07 [ disp = (disp_00_07 <<1) + ((inst_start + 4) & 0xfffffffc); ] { local tmp:4 = disp; export tmp; }

disppc4: @(disp,pc) is disp_00_07 [ disp = (disp_00_07 <<2) + ((inst_start + 4) & 0xfffffffc); ] { local tmp:4 = disp; export tmp; }
```

loudinthecloud replied:
Applied this suggestion, thanks

loudinthecloud commented:
Hi @VGKintsugi and @mumbel, thanks for the review, please check the fixes.

mumbel commented:
Looks great. Just a question of using the table or not for the disp, like I said in my original comment the manual does use labels, sometimes instead of @(disp,pc) so there nothing really wrong with it

loudinthecloud commented:
> Looks great. Just a question of using the table or not for the disp, like I said in my original comment the manual does use labels, sometimes instead of @(disp,pc) so there nothing really wrong with it

Thanks for the review @mumbel! fixed everything you mentioned and rebased the fork.

mumbel commented:
@loudinthecloud closing this.  Let me know if that pc constructor doesn't work like you had wanted, but the export __should__ produce the label you're going after (though now i'm wondering if that's only in decompiler or both)
